### PR TITLE
New loss functions, refactored engines to only implement forward_method, quality fixes

### PR DESCRIPTION
**Pull request by georgeyiasemis**
Implements:
* New loss functions (`NMSE`, `NRMSE`, `NMAE`, `SobelGradL1Loss`, `SobelGradL2Loss`)
* `mri_models` performs `_do_iteration method`, child engines perform `forward_function` which returns output_image and/or output_kspace
* Changes/Additions in `mri_transforms`
    * Padding computed as a Tensor with `ComputePadding` transform (this is helpful when cropping image and tranforming to kspace)
    * `ApplyPadding` transform
    * ComputeImage transform choices of mod output or not
    * `RenameKeys` transform
    * `Normalize` split to `ComputeScalingFactor` and `Normalize`
 * Some quality changes
 * Some documentation changes

### CODE REVIEW

File: direct/data/mri_transforms.py [hunk 1]
```diff
@@ -160,37 +149,62 @@ def __call__(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
+        ), f"Key {self.input_kspace_key} corresponding to `input_kspace_key` not found in sample."
+        input_kspace = sample[self.input_kspace_key]
+
+        assert (
+            self.sampling_mask_key in sample
+        ), f"Key {self.sampling_mask_key} corresponding to `sampling_mask_key` not found in sample."
+        sampling_mask = sample[self.sampling_mask_key]
+
+        target_kspace, _ = T.apply_mask(input_kspace, sampling_mask)
+        sample[self.target_kspace_key] = target_kspace
```
Comment (jonasteuwen) [inline direct/data/mri_transforms.py:207]: Don't use `assert` in non-tests. Raise a proper exceptions, there are already some, define a new one.

**Review by georgeyiasemis — COMMENTED** [review]
Looks good

File: direct/data/mri_transforms.py [hunk 2]
```diff
@@ -160,37 +149,62 @@ def __call__(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
-        sample["masked_kspace"] = masked_kspace
-
+        assert (
+            self.input_kspace_key in sample
+        ), f"Key {self.input_kspace_key} corresponding to `input_kspace_key` not found in sample."
+        input_kspace = sample[self.input_kspace_key]
+
+        assert (
+            self.sampling_mask_key in sample
+        ), f"Key {self.sampling_mask_key} corresponding to `sampling_mask_key` not found in sample."
```
Comment (jonasteuwen) [inline direct/data/mri_transforms.py:203]: Define a proper exception, no asserts in non-tests

File: direct/data/mri_transforms.py [hunk 3]
```diff
@@ -298,6 +312,99 @@ def __call__(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
+    r"""Computes zero padding present in multi-coil kspace input.
+
+    Zero-padding is computed from multi-coil kspace with no signal contribution, i.e. its magnitude
+    is really close to zero:
+
+    .. math ::
+
+        \text{padding} = \sum_{i=1}^{n_c} |y_i| < \frac{1}{n_x \cdot n_y}
+        \sum_{j=1}^{n_x \cdot n_y} \big\{\sum_{i=1}^{n_c} |y_i|\big\} * \epsilon.
+    """
```
Comment (jonasteuwen) [inline direct/data/mri_transforms.py:325]: Can you explain why this is needed or not available a priori?

File: direct/data/mri_transforms.py [hunk 4]
```diff
@@ -346,27 +453,31 @@ def __call__(
...
             sample[self.target_key] = T.root_sum_of_squares(image, dim=coil_dim)
-        elif self.type_reconstruction == "sense":
+        elif self.type_reconstruction in ["sense", "sense_mod"]:
             if "sensitivity_map" not in sample:
                 raise ValueError("Sensitivity map is required for SENSE reconstruction.")
             sample[self.target_key] = T.complex_multiplication(T.conjugate(sample["sensitivity_map"]), image).sum(
                 coil_dim
             )
+        if "mod" in self.type_reconstruction:
+            sample[self.target_key] = T.modulus(sample[self.target_key], complex_dim)
```
Comment (jonasteuwen) [inline direct/data/mri_transforms.py:480]: Enumerate the possibilities, don't do string matching

File: direct/functionals/__init__.py [hunk 5]
```diff
@@ -1,5 +1,9 @@
 # coding=utf-8
 # Copyright (c) DIRECT Contributors
+
 from direct.functionals.challenges import *
+from direct.functionals.grad import *
+from direct.functionals.nmae import NMAELoss
+from direct.functionals.nmse import *
```
Comment (jonasteuwen) [inline direct/functionals/__init__.py:7]: Don't use `*` imports. Unless you define `__all__` in these modules

File: direct/common/subsample.py [hunk 6]
```diff
@@ -781,19 +781,20 @@ class VariableDensityPoissonMaskFunc(BaseMaskFunc):
...
     ):
         """Inits :class:`VariableDensityPoissonMaskFunc`.
 
         Parameters
         ----------
         accelerations: list or tuple of positive numbers
             Amount of under-sampling.
-        center_scales: list or tuple of floats
+        center_fractions: list or tuple of floats
             Must have the same lenght as `accelerations`. Amount of center fully-sampling.
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/common/subsample.py]: Done

File: direct/data/mri_transforms.py [hunk 7]
```diff
@@ -318,23 +425,23 @@ def __init__(
         backward_operator: callable
             The backward operator, e.g. some form of inverse FFT (centered or uncentered).
         type_reconstruction: str
-            Type of reconstruction. Can be 'complex', 'sense' or 'rss'. Default: 'complex'.
+            Type of reconstruction. Can be "complex", "complex_mod", "sense", "sense_mod" or "rss". Default: "complex".
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/data/mri_transforms.py:428]: Used Enum

File: tests/tests_nn/test_recurrentvarnet_engine.py [hunk 8]
```diff
@@ -15,10 +15,11 @@
 
 def create_sample(shape, **kwargs):
     sample = dict()
-    sample["masked_kspace"] = torch.from_numpy(np.random.randn(*shape)).float()
+    sample["kspace"] = torch.from_numpy(np.random.randn(*shape)).float()
     sample["sensitivity_map"] = torch.from_numpy(np.random.randn(*shape)).float()
-    sample["sampling_mask"] = torch.from_numpy(np.random.randn(1, shape[1], shape[2], 1)).float()
-    sample["target"] = torch.from_numpy(np.random.randn(shape[0], shape[1], shape[2])).float()
+    sample["sampling_mask"] = torch.from_numpy(np.random.randn(shape[0], 1, shape[2], shape[3], 1)).float()
+    sample["masked_kspace"] = sample["kspace"] * sample["sampling_mask"]
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply tests/tests_nn/test_recurrentvarnet_engine.py:21]: Yeah it's in testing, doesn't really matter there as input do not really correspond to real input

File: direct/data/datasets.py [hunk 9]
```diff
@@ -451,6 +451,9 @@ def __getitem__(self, idx: int) -> Dict[str, Any]:
         num_z = kspace.shape[1]
         kspace[:, int(np.ceil(num_z * self.sampling_rate_slice_encode)) :, :] = 0.0 + 0.0 * 1j
 
+        sample["padding_left"] = 0
+        sample["padding_right"] = np.all(np.abs(kspace).sum(-1) == 0, axis=0).nonzero()[0][0]
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/data/datasets.py:455]: Yes, at least for the CC data that we have

File: direct/data/mri_transforms.py [hunk 10]
```diff
@@ -160,37 +149,62 @@ def __call__(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
+        target_kspace_key: str
+            Default "masked_kspace".
+        """
         super().__init__()
         self.logger = logging.getLogger(type(self).__name__)
 
+        self.sampling_mask_key = sampling_mask_key
+        self.input_kspace_key = input_kspace_key
+        self.target_kspace_key = target_kspace_key
+
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/data/mri_transforms.py:179]: Nope, in SSL transforms they are not ;)

File: direct/data/mri_transforms.py [hunk 11]
```diff
@@ -160,37 +149,62 @@ def __call__(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
     def __call__(self, sample: Dict[str, Any]) -> Dict[str, Any]:
         """Calls :class:`ApplyMask`.
 
-        This assumes that a `sampling_mask` is present in the sample.
+        Applies mask with key `sampling_mask_key` onto kspace `input_kspace_key`. Result is stored as a tensor with
+        key `target_kspace_key`.
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/data/mri_transforms.py:184]: Yes, in SSL settings, you might have two or more sampling masks (eg splits of the sampling_mask -> theta_sampling mask & lambda_sampling_mask)

File: direct/functionals/__init__.py (hunk 5 above)
Comment (georgeyiasemis) [reply direct/functionals/__init__.py:7]: There is an __all__ in all of those

File: direct/data/h5_data.py [hunk 12]
```diff
@@ -114,6 +114,8 @@ def __init__(
             self.logger.info("Attempting to load %s filenames.", len(filenames_filter))
             filenames = filenames_filter
 
+        filenames = [pathlib.Path(_) for _ in filenames]
+
         if regex_filter:
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/data/h5_data.py:119]: Because the filenames can be also as input from `filenames_filter`

File: direct/data/mri_transforms.py (hunk 4 above)
Comment (georgeyiasemis) [reply direct/data/mri_transforms.py:480]: Done

File: direct/data/mri_transforms.py [hunk 13]
```diff
@@ -318,23 +425,23 @@ def __init__(
...
         self.backward_operator = backward_operator
         self.kspace_key = kspace_key
         self.target_key = target_key
 
-        self.type_reconstruction = type_reconstruction
+        self.type_reconstruction = type_reconstruction.lower()
 
-        if type_reconstruction.lower() not in ["complex", "sense", "rss"]:
+        if type_reconstruction not in ["complex", "complex_mod", "sense", "sense_mod", "rss"]:
             raise ValueError(
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/data/mri_transforms.py:438]: Done

File: direct/data/mri_transforms.py [hunk 14]
```diff
@@ -346,27 +453,31 @@ def __call__(
...
-        if self.type_reconstruction == "complex":
+        if self.type_reconstruction in ["complex", "complex_mod"]:
             sample[self.target_key] = image.sum(coil_dim)
-        elif self.type_reconstruction.lower() == "rss":
+        elif self.type_reconstruction == "rss":
             sample[self.target_key] = T.root_sum_of_squares(image, dim=coil_dim)
-        elif self.type_reconstruction == "sense":
+        elif self.type_reconstruction in ["sense", "sense_mod"]:
             if "sensitivity_map" not in sample:
                 raise ValueError("Sensitivity map is required for SENSE reconstruction.")
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/data/mri_transforms.py]: Resolved

**Pull request merged**